Fix backwards instructions in left-side driving scenarios

Guardiola31337 opened the pull request:
## Description

Fixes backwards instructions in left-side driving scenarios.

Fixes #396 and a regression from https://github.com/mapbox/mapbox-navigation-android/pull/1946

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Maneuvers images (both the maneuver view and the notification maneuver icon) should be displayed accordingly in left-side driving countries.

### Implementation

Fixed `ManeuverView` regression and added new notification maneuver icons and logic for left driving side 👀 `ManeuverUtils#getManeuverResource` 

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR

cc @1ec5

andrlee commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/ManeuverUtils.java:
```diff
@@ -7,22 +7,29 @@
-        if (!STEP_MANEUVER_MODIFIER_LEFT.equals(drivingSide)) {
-          drivingSide = EMPTY;
+        if (STEP_MANEUVER_MODIFIER_LEFT.equals(drivingSide)) {
+          if (STEP_MANEUVER_MODIFIER_UTURN.equals(maneuverModifier)
```
i wonder if we could reduce number (5) of nested `if` conditions to improve readability of this method.

andrlee commented on libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/ManeuverMap.java:
```diff
@@ -141,6 +141,21 @@
       R.drawable.ic_maneuver_roundabout_straight);
     maneuverMap.put(STEP_MANEUVER_TYPE_ROUNDABOUT, R.drawable.ic_maneuver_roundabout);
 
+    maneuverMap.put(STEP_MANEUVER_TYPE_ROUNDABOUT + STEP_MANEUVER_MODIFIER_LEFT + STEP_MANEUVER_MODIFIER_LEFT,
```
what's the benefit of maintaining hashmap of the constants vs. simple `switch-case`? also note is that `ManeuverMap` constructor is not private, which may potentially result in multiple instances of the `ManeuverMap`object.

Guardiola31337 replied:
> what's the benefit of maintaining hashmap of the constants vs. simple `switch-case`?

As Martin Fowler describes in _Refactoring Improving the Design of Existing Code_:

> The problem with switch statements is essentially that of duplication. Often you
find the same switch statement scattered about a program in different places. If you add a new clause to the switch, you have to find all these switch, statements and change them. The object oriented notion of polymorphism gives you an elegant way to deal with this problem.
>
> Most times you see a switch statement you should consider polymorphism.

So in general, switch statements are considered a bad smell 👀 https://refactoring.guru/smells/switch-statements

Most of the times a  switch statement can be easily replaced with map / dictionary.

And yeah, it's true there’s probably a bit more overhead in the dictionary than in the switch statement because you have to store the dictionary somewhere so it’s obviously using some memory. However IMO, in return you’re getting clean, concise, less error-prone and well-organized code that helps maintainability in the long term.

andrlee commented on libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/maneuver/ManeuverView.java:
```diff
@@ -234,9 +234,12 @@ protected void onDraw(Canvas canvas) {
-    if (drivingSide.equals(STEP_MANEUVER_MODIFIER_RIGHT)) {
-      boolean flip = SHOULD_FLIP_MODIFIERS.contains(maneuverModifier);
-      setScaleX(flip ? -1 : 1);
+    boolean flip = SHOULD_FLIP_MODIFIERS.contains(maneuverModifier);
```
is this equivalent to:
```suggestion
    boolean flip = SHOULD_FLIP_MODIFIERS.contains(maneuverModifier);
    if (STEP_MANEUVER_MODIFIER_LEFT.equals(drivingSide) && STEP_MANEUVER_MODIFIER_UTURN.contains(maneuverModifier)) {
        setScaleX(flip ? 1 : -1);
    } else {
       setScaleX(flip ? -1 : 1);
    }
```

Guardiola31337 commented:
I'm still testing multiple scenarios in different countries, I'll report back here and let you know when this is ready for review.

Guardiola31337 commented:
Tested some of the incorrect cases reported in #396 and https://github.com/mapbox/mapbox-navigation-android/pull/1946 and so far it’s working :ok_hand: 

_Borsdane Ave Banner_

![borsdane_ave_banner](https://user-images.githubusercontent.com/1668582/64466890-517f6300-d0e2-11e9-809c-6795044470e0.png)

_Borsdane Ave Notification_

![borsdane_ave_notification](https://user-images.githubusercontent.com/1668582/64466922-6fe55e80-d0e2-11e9-8872-c51559211b7a.png)

_East Lancashire Rd Banner_

![east_lancashire_rd_banner](https://user-images.githubusercontent.com/1668582/64466927-7673d600-d0e2-11e9-9da5-b86e9ef8d94f.png)

_East Lancashire Rd Notification_

![east_lancashire_rd_notification](https://user-images.githubusercontent.com/1668582/64466957-93a8a480-d0e2-11e9-9b77-c90c9ca8c68e.png)

_Flamingo Rd Banner_

![flamingo_rd_banner](https://user-images.githubusercontent.com/1668582/64466967-a4591a80-d0e2-11e9-8b09-88e48c746c14.png)

_Flamingo Rd Notification_

![flamingo_rd_notification](https://user-images.githubusercontent.com/1668582/64466971-af13af80-d0e2-11e9-8e92-0ee1a8bf6eae.png)

_Old Kent Rd Banner_

![old_kent_rd_banner](https://user-images.githubusercontent.com/1668582/64466990-bc309e80-d0e2-11e9-98ff-36ef1cda81e5.png)

_Old Kent Rd Notification_

![old_kent_rd_notification](https://user-images.githubusercontent.com/1668582/64466996-c5ba0680-d0e2-11e9-952d-7dab63ae9e25.png)

Guardiola31337 commented:
One thing that I noticed during testing is that the source of the data that we're using for the banner and the notification is different 👀

`BannerInstructions` for `ManeuverView` https://github.com/mapbox/mapbox-navigation-android/blob/71fae584248a7b9c703fe9232bdd70fa6df7dee2/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java#L493-L502

and `LegStep` for `MapboxNavigationNotification` https://github.com/mapbox/mapbox-navigation-android/blob/71fae584248a7b9c703fe9232bdd70fa6df7dee2/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigationNotification.java#L274-L281

For consistency, we should revisit / consolidate this using the banner information in a feature PR. Refs. https://github.com/mapbox/mapbox-navigation-android/pull/1946#discussion_r288317754

Guardiola31337 commented:
IMO, objects and methods should live / be where their context is. From my experience utility classes become a drawer easily and we should treat them carefully. So I went ahead and removed `ManeuverUtils` and moved `getManeuverResource` utility method into `MapboxNavigationNotification`. Also I don't believe this class needs to be public as it's used only internally (breaking change / SemVer).

This is now ready for review @andrlee @abhishek1508 @olegzil

Christophe668 commented:
> IMO, objects and methods should live / be where their context is. From my experience utility classes become a drawer easily and we should treat them carefully. So I went ahead and removed `ManeuverUtils` and moved `getManeuverResource` utility method into `MapboxNavigationNotification`. Also I don't believe this class needs to be public as it's used only internally (breaking change / SemVer).
> 
> This is now ready for review @andrlee @abhishek1508 @olegzil

@Guardiola31337, To be honest, I was using this class to retrieve to the icon resource in order to implement our upcoming/follow up leg design.
Is this something you want to restrict/avoid?

Guardiola31337 commented:
Hey @Christophe668 thanks for your feedback.

> To be honest, I was using this class to retrieve to the icon resource in order to implement our upcoming/follow up leg design.
Is this something you want to restrict/avoid?

As commented in https://github.com/mapbox/mapbox-navigation-android/pull/2044#issuecomment-529054075 we should consolidate the data source for the banner and the notification, in favor of using the banner information. Also mentioning that the logic changed is used for generating the `MapboxNavigationNotification` icons (which still has some known issues 👀 https://github.com/mapbox/mapbox-navigation-android/issues/1952).

As you may know, all this data can be retrieved from the `RouteProgress` and used to implement your upcoming / follow up leg design. Would that work for you? Are you using these resources directly in your UI? If so, I wouldn't recommend you doing that because of the reasons above-mentioned and because it could break your UI if we decide to change the resources internally. As mentioned we don't believe this class / method needs to be public as it's used only internally. On a side note, if you still want to use this logic you can grab the code and implement it on your side, that's the beauty of open source 😅

In any case, could you talk more about your specific use case? We'd love to make sure we're not missing anything. Thanks again!
